File: day-2/first.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"strconv"
	"strings"
)

func first(scanner *bufio.Scanner) {
	hposition, depth := 0, 0

	for scanner.Scan() {
		split_text := strings.Split(scanner.Text(), " ")

		direction := split_text[0]
		number, err := strconv.Atoi(split_text[1])
		if err != nil {
			log.Fatal(err)
		}

		switch direction {
		case "up":
			depth -= number
		case "down":
			depth += number
		case "forward":
			hposition += number
		}
	}

	fmt.Println("First:", hposition*depth)

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}
}

File: day-2/nodejs/src/first.js
const parse = (coordinates, [depth, hposition]) => {
    if (!coordinates.length)
        return [depth, hposition]

    const [command, number] = coordinates[0]

    switch (command) {
        case 'up':
            return parse(coordinates.slice(1), [depth - number, hposition])

        case 'down':
            return parse(coordinates.slice(1), [depth + number, hposition])

        default:
            return parse(coordinates.slice(1), [depth, hposition + number])
    }
}

export const first = (coordinates) => {
    return parse(coordinates, [0, 0]).reduce((a, b) => a * b, 1)
}